### PR TITLE
fix: murkens abduction

### LINKED ISSUE #468: Murkens is playing my card from opponent's archive - probably should go to my hand
image

(lazy to type)

### PR DESCRIPTION
**Pull request by dmikalova**
fix: murkens and abduction - eg should return to hand - fixes #468 

---

> [!NOTE]
> **Medium Risk**
> Touches core zone-move logic (`player.moveCard`) plus discard/purge resolution to implement a new replacement effect, which can affect many card interactions if edge cases are missed. Added broad test coverage reduces risk, but the change impacts fundamental game-state transitions.
> 
> **Overview**
> Introduces a first-class `ability.actions.abduct()` and an `AbductAction` that archives a card while marking it `abducted`, with a **centralized replacement effect** that returns abducted cards to their owner’s hand whenever they would leave archives.
> 
> Updates core actions (`player.moveCard`, `DiscardCardAction`, `PurgeAction`, `PutIntoPlayAction`) to honor this replacement deterministically (by snapshotting the origin location and exposing `event.replaced`), and migrates multiple cards (e.g. Mass Abduction / Uxlyx / Sample Collection / Suspended Animation / Curse of Disappearances) to use `abduct` where card text requires the “owner’s hand instead” behavior.
> 
> Adjusts a few related card flows to account for replacement outcomes (e.g. Yzphyz Knowdrone follow-up only if purge wasn’t replaced; Parallel Thinking steal condition ignores replaced discards; Crop Circles capture counts actual purges), expands the test harness (`keys` setter, `purged` getter), and adds extensive regression tests for abduction interactions (including Murkens, gigantics, tokens, and purge/discard effects).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6793c00c8d033575c01ab42263e60e31b29b8088. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: server/game/player.js [hunk 1]
```diff
@@ -524,6 +524,24 @@ class Player extends GameObject {
             card.purgedBy = null;
         }
 
+        // Abduct replacement effect: abducted cards go to their owner's hand
+        // instead of anywhere else when leaving archives
+        if (location === 'archives' && card.abducted) {
+            if (targetLocation !== 'hand') {
+                this.game.addMessage(
```
Comment (strib) [inline server/game/player.js]: Isn't this a duplicate message to the one printed in the discard/purge actions?

File: server/game/GameActions/DiscardCardAction.js [hunk 2]
```diff
@@ -106,13 +106,34 @@ class DiscardCardAction extends CardGameAction {
...
+            () => {
+                // Abduct replacement effect: abducted cards go to owner's hand instead of discard
+                if (location === 'archives' && card.abducted) {
+                    context.game.addMessage(
+                        "{0} leaves {1}'s archives and is added to {2}'s hand instead of being discarded",
+                        card,
+                        card.controller,
+                        card.owner
+                    );
+                    card.abducted = false;
```
Comment (strib) [inline server/game/GameActions/DiscardCardAction.js]: This is done by `moveCard()`, right?  (For purge also.)

**Comment by dmikalova** [comment]
Put into draft until we have the message tests merged

File: server/game/GameActions/DiscardCardAction.js (hunk 2 above)
Comment (dmikalova) [reply server/game/GameActions/DiscardCardAction.js]: Updated this to happen centrally in moveCard

File: server/game/player.js (hunk 1 above)
Comment (dmikalova) [reply server/game/player.js]: It is a duplicate message, but I think it adds clarity to say "x discards y" and the next line say "... instead of being discarded". It's also a lot easier to have this kind of duplicate in the current messaging system.

> 2026-05-08 23:27:43 player1 uses Uxlyx the Zookeeper to reap with Uxlyx the Zookeeper
2026-05-08 23:27:43 player1 uses Uxlyx the Zookeeper to abduct Batdrone
2026-05-08 23:27:43 player1 plays Harmonic Pack
2026-05-08 23:27:43 player1 gains an amber due to Harmonic Pack's bonus icon
2026-05-08 23:27:43 player1 chooses option 'Mine'
2026-05-08 23:27:43 player1 uses Harmonic Pack to deal 2 damage to Krump
2026-05-08 23:27:43 Harmonic Pack reveals Batdrone
2026-05-08 23:27:43 player1 uses Harmonic Pack to deal 3 more damage to Krump and discard Batdrone
2026-05-08 23:27:43 player1 uses Harmonic Pack to discard Batdrone
2026-05-08 23:27:43 Batdrone leaves player1's archives and is added to player2's hand instead of being discarded

File: server/game/GameActions/DiscardCardAction.js [hunk 3]
```diff
@@ -106,13 +106,27 @@ class DiscardCardAction extends CardGameAction {
...
+                // Abduct replacement: delegate to moveCard, which redirects
+                // archived+abducted cards to their owner's hand and emits the
+                // "...instead of being discarded" message.
+                if (location === 'archives' && card.abducted) {
+                    card.controller.moveCard(card, 'discard');
+                    event.replaced = true;
+                    return;
+                }
+
+                if (card.location === 'hand') {
```
Comment (strib) [inline server/game/GameActions/DiscardCardAction.js:131]: Maybe this should be `location` too?  (or make the above check `card.location`?)  Not sure if there's a difference.

File: server/game/GameActions/PurgeAction.js [hunk 4]
```diff
@@ -89,15 +89,30 @@ class PurgeAction extends CardGameAction {
...
+                // Abduct replacement: delegate to moveCard, which redirects
+                // archived+abducted cards to their owner's hand and emits the
+                // "...instead of being purged" message.
+                if (location === 'archives' && card.abducted) {
+                    card.controller.moveCard(card, 'purged');
+                    event.replaced = true;
+                    return;
+                }
+
+                if (card.location === 'play area') {
```
Comment (strib) [inline server/game/GameActions/PurgeAction.js:115]: Same.

**Review by strib — APPROVED** [review]
Great cleanup, thanks!

File: server/game/GameActions/DiscardCardAction.js (hunk 3 above)
Comment (dmikalova) [reply server/game/GameActions/DiscardCardAction.js:131]: This is intentional - abduct should ignore any changes by other abilities and move to hand, whereas discard counting should use the current location to prevent double counting. Added comments to this effect.

**Review by strib — APPROVED** [review]
